### PR TITLE
Fix `VotingPowerCalculator::voting_power_in`

### LINKED ISSUE #282: `BlockIdFlagNil` votes counted when verifying a commit
Comparing the implementation of Light Client verification with the Go one, there is a difference in the calculation of the voting power for a commit.

Methods `verify_commit_full` and `verify_commit_trusting` in `lite/verifier.rs` use a method [`voting_power_in`](https://github.com/informalsystems/tendermint-rs/blob/afb11bf21fc3899edaf157b4f95e1b94a158215b/tendermint/src/lite_impl/signed_header.rs#L17) which sums up voting power for all validators in the commit which have a non absent vote.

Comparing this to the Go implementation in functions `VerifyCommit` and `VerifyCommitTrusting` in `types/validator_set.go` where voting power is counted by summing up only those which have a [`BlockIdFlagCommit`](https://github.com/tendermint/tendermint/blob/master/types/validator_set.go#L681) vote.

This then results in Rust implementation accepting commits which have no `BlockIdFlagCommit` votes but only `BlockIdFlagNil`.

### LINKED ISSUE #335: Don't add voting power of the same validator multiple times in voting_power_in
There is a TODO already in the code: https://github.com/informalsystems/tendermint-rs/blob/184a1f9925f541473313efcd483c32cb913f8af9/light-client/src/operations/voting_power.rs#L38

### PR DESCRIPTION
**Pull request by romac**
Fix #282 
Fix #335

 - Provide two helper methods for checking validators overlap and if there is enough trust w.r.t. the trusted state
- Only count `Commit` signatures towards the voting power (#282)
- Do not rely on `signed_votes` method to get the votes
- Only count the voting power of a validator once (#335)

### Notes

- [x] Needs to throw a proper error in some places
- [x] Trust threshold is currently hardcoded and should be rather passed in
- [x] The method should likely be renamed or its signature should be changed to reflect the new behavior
- [ ] ~~Most of logic and comments are taken from the Go code, so perhaps some attribution should be added as well~~
- [ ] ~~Tests are failing because the signatures currently fail to verify, likely because of mismatch with their validator~~

---

* [x] Referenced an issue explaining the need for the change
* [x] Updated all relevant documentation in docs
* [x] Updated all code comments where relevant
* [ ] Wrote tests
* [ ] Updated CHANGES.md

### CODE REVIEW

File: light-client/src/operations/voting_power.rs [hunk 1]
```diff
@@ -50,36 +53,104 @@ impl VotingPowerCalculator for ProdVotingPowerCalculator {
...
+        for (idx, signature) in signatures.into_iter().enumerate() {
+            if signature.is_absent() {
+                continue; // OK, some signatures can be absent.
+            }
+
+            // The vals and commit have a 1-to-1 correspondance (see check above).
+            // This means we don't need the validator address or to do any lookup.
+            let val = validators[idx];
 
-        for vote in &signed_header.signed_votes() {
```
Comment (Liamsi) [inline light-client/src/operations/voting_power.rs:59]: Thanks for doing this @romac 🙇 Can't we then delete the `signed_votes()` method?

Comment (romac) [reply light-client/src/operations/voting_power.rs:59]: I guess we could, but since this is very much a WIP I didn't remove any code yet.

File: light-client/tests/light_client.rs [hunk 2]
```diff
@@ -152,13 +153,16 @@ fn run_bisection_test(tc: TestBisection<LightBlock>) {
     let trusting_period = tc.trust_options.period;
     let now = tc.now;
 
+    // FIXME: What should this be, and where should it be configured?
```
Comment (melekes) [inline light-client/tests/light_client.rs:156]: in Go, default is 10sec. once we switch to the proposer based timestamps, it will probably be a consensus parameter

File: light-client/src/operations/voting_power.rs [hunk 3]
```diff
@@ -50,36 +53,104 @@ impl VotingPowerCalculator for ProdVotingPowerCalculator {
...
     }
 }
+
+fn vote_from_non_absent_signature(
+    commit_sig: &CommitSig,
+    validator_index: u64,
+    commit: &Commit,
+) -> Option<Vote> {
+    let (validator_address, timestamp, signature, block_id) = match commit_sig {
+        CommitSig::BlockIDFlagAbsent { .. } => return None,
```
[the remark this answers is not in the document]
Comment (romac) [reply light-client/src/operations/voting_power.rs:158]: This function is actually the one doing the filtering for non absent signatures, hence the need for the Option. I guess the name isn't great, I am open to suggestions :)

**Comment by romac** [comment]
I pushed a new version which should now be doing the right thing w.r.t. checking trust vs validators signers, but still relies on validator addresses. I guess we can address the latter in a future PR once we are convinced that this version is correct.

File: light-client/src/operations/voting_power.rs [hunk 4]
```diff
@@ -1,19 +1,78 @@
...
     fn total_power_of(&self, validators: &ValidatorSet) -> u64;
-    fn voting_power_in(
+
+    fn check_enough_trust(
+        &self,
+        untrusted_header: &SignedHeader,
+        untrusted_validators: &ValidatorSet,
+        trust_threshold: TrustThreshold,
+    ) -> Result<VotingPower, VerificationError> {
+        println!("check_validators_overlap");
```
Comment (brapse) [inline light-client/src/operations/voting_power.rs:42]: This should probably be removed.

File: light-client/src/operations/voting_power.rs [hunk 5]
```diff
@@ -1,19 +1,78 @@
...
+            Err(VerificationError::NotEnoughTrust(voting_power))
+        }
+    }
+
+    fn check_signers_overlap(
+        &self,
+        untrusted_header: &SignedHeader,
+        untrusted_validators: &ValidatorSet,
+    ) -> Result<VotingPower, VerificationError> {
+        println!("check_signers_overlap");
```
Comment (brapse) [inline light-client/src/operations/voting_power.rs:58]: Can probably be removed.

File: light-client/src/operations/voting_power.rs [hunk 6]
```diff
@@ -1,19 +1,78 @@
...
+        let two_thirds = TrustThreshold::new(2, 3).unwrap();
```
Comment (brapse) [inline light-client/src/operations/voting_power.rs:59]: Should this perhaps be a package level constant?

**Comment by romac** [comment]
It would be great if someone could run the tests of the `light-client` crate and inspect the output to ensure that errors that are raised are indeed the ones we expect from the JSON file. I did so myself 

```
$ cd light-client/
$ cargo test -- --nocapture --test-threads 1
```

I already did so myself, and I am not too sure about this test:

```
Running light client against bisection test-file: ./tests/support/bisection/single_peer/not_enough_commits.json
  - Case: Trusted height=1, fails at height 6 because more than one-third (trust level) vals didn't sign
[light-client/tests/light_client.rs:247] e = Error(
    Context {
        kind: InvalidLightBlock(
            InvalidNextValidatorSet {
                header_next_validators_hash: Hash::Sha256(75B9F27F25F3515EF7CA61DD42C272CCC280C537E6A9ED493710EB361CB90B0D),
                next_validators_hash: Hash::Sha256(26952B5D784A1564D167DF98D2D37376B5E77771928256D25E6FF9AE3AD11564),
            },
        ),
...
```

This test is passing because we expect an error, but I don't think that that is the right one, except if there's perhaps more than one error encoded in the JSON test. In particular, I would have expected  the test to result in either `NotEnoughTrust` or `InsufficientValidatorsOverlap` but we are getting `InvalidNextValidatorSet` instead. 

@Shivani912 Is that something you could look into?

File: light-client/src/operations/voting_power.rs [hunk 7]
```diff
@@ -24,39 +83,111 @@ impl VotingPowerCalculator for ProdVotingPowerCalculator {
...
-                None => continue,
+        validator_set: &ValidatorSet,
+        trust_threshold: TrustThreshold,
+    ) -> Result<VotingPower, VerificationError> {
+        let signatures = &signed_header.commit.signatures;
+
+        let mut tallied_voting_power = 0_u64;
+        let mut seen_validators = HashSet::new();
+
+        for (idx, signature) in signatures.into_iter().enumerate() {
```
Comment (brapse) [inline light-client/src/operations/voting_power.rs:97]: Does it make sense to reference the test vector which exercises this particular edge case?

File: light-client/src/operations/voting_power.rs [hunk 8]
```diff
@@ -24,39 +83,111 @@ impl VotingPowerCalculator for ProdVotingPowerCalculator {
...
+            // If the vote is neither absent nor nil, tally its power
+            if signature.is_commit() {
+                tallied_voting_power += validator.power();
+            } else {
+                // It's OK. We include stray signatures (~votes for nil)
+                // to measure validator availability.
             }
 
-            signed_power += val.power();
+            // TODO: Break out when we have enough voting power
```
Comment (brapse) [inline light-client/src/operations/voting_power.rs:140]: Reference #235

File: light-client/src/operations/voting_power.rs (hunk 1 above)
Comment (romac) [reply light-client/src/operations/voting_power.rs:36]: Will do this as part of #342.

**Comment by Shivani912** [comment]
> It would be great if someone could run the tests of the `light-client` crate and inspect the output to ensure that errors that are raised are indeed the ones we expect from the JSON file. I did so myself
> 
> ```
> $ cd light-client/
> $ cargo test -- --nocapture --test-threads 1
> ```
> 
> I already did so myself, and I am not too sure about this test:
> 
> ```
> Running light client against bisection test-file: ./tests/support/bisection/single_peer/not_enough_commits.json
>   - Case: Trusted height=1, fails at height 6 because more than one-third (trust level) vals didn't sign
> [light-client/tests/light_client.rs:247] e = Error(
>     Context {
>         kind: InvalidLightBlock(
>             InvalidNextValidatorSet {
>                 header_next_validators_hash: Hash::Sha256(75B9F27F25F3515EF7CA61DD42C272CCC280C537E6A9ED493710EB361CB90B0D),
>                 next_validators_hash: Hash::Sha256(26952B5D784A1564D167DF98D2D37376B5E77771928256D25E6FF9AE3AD11564),
>             },
>         ),
> ...
> ```
> 
> This test is passing because we expect an error, but I don't think that that is the right one, except if there's perhaps more than one error encoded in the JSON test. In particular, I would have expected the test to result in either `NotEnoughTrust` or `InsufficientValidatorsOverlap` but we are getting `InvalidNextValidatorSet` instead.
> 
> @Shivani912 Is that something you could look into?

@romac There's a bug in this one (issue #290) I'm working on it and will do a PR once I fix it! 

I feel we really need to have checks (in the test itself) on errors even if the tests pass. I tried to do this at the end of Q1/beginning of Q2 but it seemed it was a little too early to be able to do so. Think this should be on priority for next quarter?

File: light-client/src/operations/voting_power.rs [hunk 9]
```diff
@@ -1,19 +1,75 @@
...
+    fn check_enough_trust(
+        &self,
+        untrusted_header: &SignedHeader,
+        untrusted_validators: &ValidatorSet,
+        trust_threshold: TrustThreshold,
+    ) -> Result<VotingPower, VerificationError> {
+        let voting_power =
+            self.voting_power_of(untrusted_header, untrusted_validators, trust_threshold)?;
+
+        if trust_threshold.is_enough_power(voting_power.tallied, voting_power.total) {
```
Comment (OStevan) [inline light-client/src/operations/voting_power.rs:45]: Mentioned in one of the comments above, I would probably make more sense to just check this in the `voting_power_of` and instead of a value just return a boolean value or something else to tell us if everything is fine or there is some error.

File: light-client/src/operations/voting_power.rs [hunk 10]
```diff
@@ -1,19 +1,75 @@
...
+            self.total, self.tallied, self.trust_threshold
+        )
+    }
+}
 
 pub trait VotingPowerCalculator: Send {
     fn total_power_of(&self, validators: &ValidatorSet) -> u64;
-    fn voting_power_in(
+
+    fn check_enough_trust(
```
Comment (OStevan) [inline light-client/src/operations/voting_power.rs:36]: The name and the result of this function were for me a bit misleading as the name would suggest a boolean value (or in this case `Ok()` or `Error(...)`) and it is more or less used as such since the result value in the Ok case is just ignored.

File: light-client/src/operations/voting_power.rs [hunk 11]
```diff
@@ -1,19 +1,75 @@
...
+use std::collections::HashSet;
+use std::fmt;
+
+use tendermint::block::CommitSig;
+use tendermint::lite::types::TrustThreshold as _;
 use tendermint::lite::types::ValidatorSet as _;
+use tendermint::vote::{SignedVote, Vote};
+
+#[derive(Copy, Clone, Debug, PartialEq, Serialize, Deserialize)]
+pub struct VotingPower {
```
Comment (OStevan) [inline light-client/src/operations/voting_power.rs:17]: For me this is a bit misleading as the struct has more information than what I would expect by reading just the name. Also, I think it would be beneficial to implement the TODO about the breaking out of the loop as than this would go probably go away.

File: light-client/src/operations/voting_power.rs [hunk 12]
```diff
@@ -24,39 +80,107 @@ impl VotingPowerCalculator for ProdVotingPowerCalculator {
...
+                Some(vote) => vote,
+                None => continue, // Ok, some signatures can be absent
+            };
+
+            // Ensure we only count a validator's power once
+            if seen_validators.contains(&vote.validator_address) {
+                continue;
+            } else {
+                seen_validators.insert(vote.validator_address);
+            }
```
Comment (OStevan) [inline light-client/src/operations/voting_power.rs:109]: One interesting difference between go code and the implementation here is this `if`. (go) In case of 2/3 threshold this check is done by comparing the size of the signers validator set and the expected validator set, if both are the same length everything is fine and we don't even do these checks. However in the case where we are interested in trust between two non adjacent headers, `then` branch should raise an exception.

File: light-client/src/operations/voting_power.rs [hunk 13]
```diff
@@ -24,39 +83,111 @@ impl VotingPowerCalculator for ProdVotingPowerCalculator {
...
+        validator_set: &ValidatorSet,
+        trust_threshold: TrustThreshold,
+    ) -> Result<VotingPower, VerificationError> {
+        let signatures = &signed_header.commit.signatures;
+
+        let mut tallied_voting_power = 0_u64;
+        let mut seen_validators = HashSet::new();
+
+        for (idx, signature) in signatures.into_iter().enumerate() {
+            let vote = vote_from_non_absent_signature(signature, idx as u64, &signed_header.commit);
```
[the remark this answers is not in the document]
Comment (romac) [reply light-client/src/operations/voting_power.rs]: You mean counting absent signatures? I don't think that's possible since a `CommitSig::BlockIDFlagAbsent` does not contain any of the data required to craft a `Vote` (ie. validator_address, timestamp and signature).

File: light-client/src/operations/voting_power.rs (hunk 7 above)
Comment (romac) [reply light-client/src/operations/voting_power.rs]: What edge case are you referring to here?

**Comment by romac** [comment]
> @romac There's a bug in this one (issue #290) I'm working on it and will do a PR once I fix it!

Awesome, thanks :)

> I feel we really need to have checks (in the test itself) on errors even if the tests pass. I tried to do this at the end of Q1/beginning of Q2 but it seemed it was a little too early to be able to do so. Think this should be on priority for next quarter?

Yes, that would definitely increase our confidence in the tests and thus the correctness of implementation :)

File: light-client/src/operations/voting_power.rs (hunk 11 above)
Comment (romac) [reply light-client/src/operations/voting_power.rs:17]: Agreed that the name is perhaps not the best, but we still need this information to be able to yield a meaningful error in case of lack of trust/signers overlap.

File: light-client/src/operations/voting_power.rs (hunk 12 above)
Comment (romac) [reply light-client/src/operations/voting_power.rs:110]: @melekes What do you think?

Comment (romac) [reply light-client/src/operations/voting_power.rs:110]: I guess raising an error is probably the proper behavior there. Will already change that, and we can always revert it back if that was wrong.

File: light-client/src/operations/voting_power.rs (hunk 7 above)
Comment (brapse) [reply light-client/src/operations/voting_power.rs]: Oops perhaps I did not comment the correct line and actually there is an informative comment further down `Ensure we only count a validator's power once`

File: light-client/src/operations/voting_power.rs (hunk 13 above)
[the remark this answers is not in the document]
Comment (brapse) [reply light-client/src/operations/voting_power.rs]: True!

File: light-client/src/operations/voting_power.rs (hunk 7 above)
Comment (romac) [reply light-client/src/operations/voting_power.rs]: Unfortunately, we currently do have a test vector which covers this edge case.

cc @Shivani912

File: light-client/src/operations/voting_power.rs [hunk 14]
```diff
@@ -1,19 +1,76 @@
...
+    }
+}
 
 pub trait VotingPowerCalculator: Send {
     fn total_power_of(&self, validators: &ValidatorSet) -> u64;
+
+    fn check_enough_trust(
+        &self,
+        untrusted_header: &SignedHeader,
+        untrusted_validators: &ValidatorSet,
```
Comment (ebuchman) [inline light-client/src/operations/voting_power.rs:39]: These are trusted validators

**Comment by romac** [comment]
@ebuchman 

> I think we still want to break the voting_power_in into two functions, one for check_signers and one for check_enough_trust

It would be great if you could describe the differences between the two versions somewhere.

> and neither should depend on vote.validator_address

I initially tried to fix that as well but eventually chose to focus on fixing the computation itself, and improve it later.

**Comment by ebuchman** [comment]
Done in https://github.com/informalsystems/tendermint-rs/issues/377
